release-22.1: backupccl: elide spans from backups that were subsequently reintroduced

msbutler opened the pull request:
Backport 1/1 commits from #87312.

/cc @cockroachdb/release

---

Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See https://github.com/cockroachdb/cockroach/issues/87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (https://github.com/cockroachdb/cockroach/issues/88042, https://github.com/cockroachdb/cockroach/issues/88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs https://github.com/cockroachdb/cockroach/issues/87305

Release justification: bug fix

Release note: fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198; https://www.cockroachlabs.com/docs/advisories/a88042

msbutler committed:
backupccl: elide spans from backups that were subsequently reintroduced

Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See #87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (#88042, #88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs #87305

Release justification: bug fix

Release note (bug fix): fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88488)
<!-- Reviewable:end -->

msbutler committed:
backupccl: reintroduce previously offline tables with manifest.DescriptorChanges

getReintroducedSpans finds all tables included in the current and previous
backup that may have undergone a non-mvcc operation. The current backup will
then back up these tables' spans from ts = 0, as the previous backup may have
missed certain non-mvcc written.

To find these tables, getReintroducedSpans must find all tables covered in the
previous backup that were in the offline state at previous backup start time.

This function assumed that the previous backup's
manifest.Descriptors field would contain all tables covered in the previous
backup; however, while investigating #88042, we discovered that this assumption
is not correct. During revision history backups, a table with an in-progress
import (e.g. offline at backup time) can get backed up and included in
manifest.DescriptorChanges but not in manifest.Descriptors. This implies that
getReintroducedSpans missed reintroducing spans from this table, implying that
backup chains have missed backing up some data.

This patch fixes getReintroducedSpans to ensure it reintroduces tables included
in manifest.DescriptorChanges whose last revision brought the table offline. In
addition, this patch adds a check to restore planning which causes the restore
to fail immedidiately if any restore target is missing an introduced span.

Release note(bug fix): fix apart of Tech Advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

Release justification: bug fix

msbutler commented:
Note that the 22.1 backport required the following additional changes:
 - adding the MakeFrontierAt function
 - an additional commit in to address https://github.com/cockroachdb/cockroach/issues/88042 . I'll also need to forward port this to 22.2 in order to handle cross version backup chains, but it's easier to test in 22.1, so I added it here first.
 - a few test-only changes that I plan to also change in master / 22.2

msbutler commented:
RFAL! I'd review commit by commit. As discussed over meets:
1) y'all have never seen the first commit. I plan to forward port it to 22.2.
2) the second commit is the same as what's on 22.2, aside from a a few function signature and test only changes. I also had to add the MakeFrontierAt function.

adityamaru commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
+			}
+		}
+
+		// Gather the introduced tables
```
nit: Gather the tables that were reintroduced in the current backup (i.e. backed up from ts=0).

msbutler replied:
Done.

adityamaru commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
+				if !inPrevBackup && !introduced {
+					tableError := errors.Newf("table %v was not properly backed up after its import"+
+						" completed. To continue the restore, "+
+						"either 1) restore from a different full backup; 2) restore to a system time before"+
```
Should we say something like:

"table %v was not properly backed up because of <issue_num>. To restore this table one can:

1) Restore from the full backup.
2) Restore AOST <= %s, (I think we can get the actual timestamp from the prevBackup manifest that applies to this table?). Maybe we can stash it in `inPrevBackup` as the value. IIUC that is the last valid backup for this table.

Personally, I'd drop 3) but maybe @dt can chime in here.

msbutler replied:
I like this: 

> table %v was not properly backed up because of <issue_num>.

I don't quite understand your change to suggestion 1. My original intention in suggestion 1 is that any other backup chain in the collection will not have this error.

For 2: The latest possible timestamp is right before the import ended, which is info we cannot easily retrieve in this function. I think it would require iterating over manifest.DescriptorChanges, finding the descriptor change that corresponds to turning the table back online, and returning descChange.ModificationTime - someDelta to the user. I could also return the end time of the last incremental backup that observed the table offline (much easier).

For 3: I could see a use case for 3. suppose the user doesn't actually need the corrupt table in the database restore, but really wants to restore their data to that AOST.

adityamaru replied:
> 1. My original intention in suggestion 1 is that any other backup chain in the collection will not have this error.

Even the full backup and some incrementals of this errant chain are restoreable iiuc? Maybe we don't need point 1 if we make 2 clear enough. For point 2 isn't the last valid backup for this table the incremental before the incremental in which this table appeared online but we forgot to reintroduce it? In that case can we just take the last incrementals EndTime and say any ts before this is valid.

msbutler replied:
resolved offline.

adityamaru commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
+						"either 1) restore from a different full backup; 2) restore to a system time before"+
+						" the import completed; 3) or remove %v from the restore targets.",
+						table.Name, table.Name)
+					return errors.WithIssueLink(tableError, errors.IssueLink{
```
Lets do:

errors.WithIssueLink(errors.WithHint(....), ....) and add your steps to restore this table as a hint instead?

Also I don't feel great about saying: `A backup chain with revision history can lose data on a table with an import`.

I'd qualify the statement a little more: An incremental database backup with revision history can incorrectly backup data for a table that was running an IMPORT at the time of the previous incremental in this chain of backups.

msbutler replied:
I think using `errors.WithIssueLink(errors.WithHint(....), ....)` would look kinda weird. As my data driven test indicates below, errors.WIthIssueLink already issues a `HINT` prefix. So i think your suggestion would create two hints, which seems confusing.

adityamaru replied:
The extra `HINT` might be an artifact of what we're doing here - https://github.com/cockroachdb/cockroach/blob/fe4f495d71394647ace18c475c694a7f82ba5a61/pkg/ccl/backupccl/datadriven_test.go#L584. I can try generating a corrupt backup and then restoring it with your patch and seeing how it looks in the terminal.

msbutler replied:
aside: why change how the error msg presents in the dd test?

msbutler replied:
Not worrying about this for now.

adityamaru reviewed (changes requested):
Only reviewed comment 1, flushing it out. Overall the change looks pretty clean, nice!

msbutler commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1387,8 +1387,23 @@ func getReintroducedSpans(
+	//
+	// 2. manifest.DescriptorChanges contains a list of descriptor changes tracked
+	// in the backup. While investigating #88042, it was discovered that
+	// during revision history backups, a table can get included in
```
Done.

msbutler commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1397,6 +1412,33 @@ func getReintroducedSpans(
+		induceIntroducedSpansBug = knobs.SkipDescriptorChangeIntroduction
+	}
+
+	if induceIntroducedSpansBug {
```
Done.

msbutler commented on pkg/ccl/backupccl/datadriven_test.go:
```diff
@@ -126,6 +130,7 @@ type serverCfg struct {
 	splits                int
 	ioConf                base.ExternalIODirConfig
 	localities            string
+	customTestingKnobs    string
```
leaving this for now. 22.2 datadriven has better infra for this.

msbutler commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
+// not, then that table must be covered by the nth backup's
+// manifest.IntroducedSpans field. This check only applies to backups with a
+// start time that is less than the restore AOST.
+func checkMissingIntroducedSpans(
```
Done.

msbutler commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
+		}
+
+		if i == 0 {
+			// The full backup does not contain reintroduced spans
```
Done.

msbutler commented on pkg/sql/exec_util.go:
```diff
@@ -1567,6 +1567,11 @@ type BackupRestoreTestingKnobs struct {
 	// specified here.
 	BackupMemMonitor *mon.BytesMonitor
+
+	// SkipDescriptorChangeIntroduction simulates the bug in #88042 and ensures the
```
Done.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data:
```diff
@@ -0,0 +1,368 @@
+# The exact timeline that induces the bug is the following:
+#
+# - t0: begin import jobs and pause it
+# - t1: table or database backup with revision history
```
nit: wildcard table right, table alone won't be able to resolve the descriptor in the offline state

adityamaru replied:
Also do we have table backups in the test below?

msbutler replied:
nah. Will add.

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1387,8 +1387,23 @@ func getReintroducedSpans(
+	// were offline when the last backup was taken. To create this map, we must
+	// iterate two fields in the _last_ backup's manifest:
+	//
+	// 1. manifest.Descriptors contains a list of descriptors _explicitly_
```
friendly ping on this in case you missed it

msbutler replied:
Super weird. Didn't initially see this bc It didn't appear in the "File Changed" view. Done now.

adityamaru reviewed (changes requested):
Nice, only one blocking comment about the logic regarding picking the latest revision instead of checking all revisions. After that I'm an LGTM. I'm going to generate a corrupt backup and see how the RESTORE error message looks in the SQL shell.

msbutler commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1397,6 +1412,33 @@ func getReintroducedSpans(
+				latestTableDescChangeInLastBackup[table.GetID()] = table
+			}
+		}
+	}
```
The goal is to find all descriptors that were offline when the last backup was taken (i.e. at the backup's EndTime), according to the backup's DescriptorChanges field, i.e. descriptors that may not appear in the previous backups manifest.Descriptors field. To identify these descriptors, it's sufficient to check that the latest descriptor change for the descriptor brought the desc offline.

 In the case you describe, those descriptors _would_ appear in the manifest.Descriptors field, as it was online at prevBackup.Endtime. 
 
 I've added a clarifying comment to explain this. Lemme know if it's sufficient.

stevendanna reviewed (approved):
LGTM. Please rebase away the nit commit before merge.

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data:
```diff
@@ -0,0 +1,368 @@
+
+exec-sql
+SET CLUSTER SETTING kv.bulkio.write_metadata_sst.enabled = false;
+----
```
Heh. From the fix for master. Removed.

msbutler commented:
apologies for the re-pushing. Realized the dd tests ended up in the first commit. They don't pass unless they are in the second commit.
